### PR TITLE
Support `AbortController`

### LINKED ISSUE #449: Support `AbortController`
Node.js is adopting it for everything, so I think we should support it too.

https://developer.mozilla.org/en-US/docs/Web/API/AbortController

Node.js just added support for it in `.spawn()`: https://nodejs.org/en/blog/release/v15.5.0/

This would not replace the current `.cancel()` function, in the near-term at least.

### PR DESCRIPTION
**Pull request by jopemachine**
Fixes #449.

I'd appreciate it if you could let me know if I need to write more tests or documents.

Thanks for all your efforts for this awesome library :)

### CODE REVIEW

**Comment by ehmicky** [comment]
Thanks a lot for this contribution @jopemachine!

We might want to name this option `signal` instead of `abortSignal` to match the name currently used by `child_process.spawn()`.

More fundamentally, I am wondering whether the core `signal` option should just be used instead? I have written some potential ideas in [the issue](https://github.com/sindresorhus/execa/issues/449#issuecomment-1030065002), please let me know what you think.

**Comment by jopemachine** [comment]
> We might want to name this option signal instead of abortSignal to match the name currently used by child_process.spawn().

Thanks for pointing it out, I just updated the option.

File: index.d.ts
Comment (sindresorhus) [inline index.d.ts:224]: The same docs should be in the readme too.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.d.ts
Comment (jopemachine) [reply index.d.ts:224]: Could we add an example [here](https://github.com/sindresorhus/execa/pull/490/commits/7ed6580e7ed657be1d7e16349ec85f17777ba2e0#diff-5a831ea67cf5cf8703b0de46901ab25bd191f56b320053be9332d9a3b0d01d15R567) link to avoid document duplication?

File: readme.md
Comment (sindresorhus) [inline readme.md:114]: Move this to the `signal` option.

**Comment by jopemachine** [comment]
Should I delete [this document](https://github.com/sindresorhus/execa/blob/29bc38e6ed2ce8add74baa5a06555f97edd43a14/readme.md#cancel) too? or should I mark the cancel method as `deprecated`?

File: index.js
Comment (ehmicky) [inline index.js:113]: When the `signal` option is passed to `child_process.spawn()`, `spawned.kill()` is already automatically called by Node.js when the `abort` event is emitted.

`spawned.cancel()` also sets `error.isCanceled: true`. However, we could set it instead by replacing the following line:

https://github.com/sindresorhus/execa/blob/e4b929547ade335b0dd9d7afe1a35989e19fbc8a/index.js#L130

By:

```js
    isCanceled: parsed.options.signal && parsed.options.signal.aborted,
```

File: readme.md
Comment (ehmicky) [inline readme.md:532]: Could you please move this example to the `Cancelling a spawned process` section above? For consistency, since all examples are currently grouped in the "Usage" and "Tips" sections. Thanks!

File: readme.md
Comment (ehmicky) [inline readme.md:96]: We might want to keep the following from this example:
  - Printing `isCanceled`
  - Showing that aborts make the process rejects its returned promise
  - Calling `abort()` inside `setTimeout()`, since this mimics how it would be used in real use cases (calling it right after the process has started is unlikely)

File: index.d.ts
Comment (ehmicky) [inline index.d.ts:207]: Should we add some test for this new type in `index.test-d.ts`?

File: test/kill.js
Comment (ehmicky) [inline test/kill.js:266]: Nice tests, they look great!

File: readme.md
Comment (ehmicky) [inline readme.md:82]: We should also remove the documentation [of the `cancel()` method](https://github.com/sindresorhus/execa/blob/main/readme.md#cancel).

**Review by ehmicky — CHANGES_REQUESTED** [review]
I believe we can directly delete that section of the documentation, good catch! :+1: 

Thanks again for this PR @jopemachine! There are a couple of things I have noted. The tests look great!

**Comment by jopemachine** [comment]
> I believe we can directly delete that section of the documentation, good catch! 👍
> 
> Thanks again for this PR @jopemachine! There are a couple of things I have noted. The tests look great! (But the CI tests are currently failing)

I appreciate your detailed feedback :)
I tried to reflect on all the feedback as much as I can.
If I'm missing something or got something wrong, I appreciate your letting me know.

File: index.js
Comment (ehmicky) [inline index.js:150]: `isCanceled` is always `false` when the process is successful.

File: readme.md
Comment (ehmicky) [inline readme.md:550]:
```suggestion
When `AbortController.abort()` is called, [isCanceled](#iscanceled) becomes `false`.
```

File: index.js
Comment (ehmicky) [inline index.js:112]:
```suggestion
	const isCanceled = () => context.isCanceled || (parsed.options.signal && parsed.options.signal.aborted);
```

File: test/kill.js
Comment (ehmicky) [inline test/kill.js:266]: Those tests fail on older Node.js versions due to `AbortController` not being supported. They should be put inside a condition, maybe something like this?

```js
if (globalThis.AbortController !== undefined) {
```

**Comment by ehmicky** [comment]
Thanks a lot for those edits, they look great! And the tests are now passing on Node 16 :tada: 
I have a few more comments, but we're getting there. Thanks for your help @jopemachine.

File: index.js
Comment (jopemachine) [reply index.js:112]: It seems that when `signal` option is not using and context.isCanceled is false, 
isCanceled is `undefined` (It should be false in this case) (`false || undefined === undefined`)
Would it be better to set `signal` option's default value to false to handle this?

Comment (ehmicky) [reply index.js:112]: Good point! :+1:

File: index.js
Comment (ehmicky) [inline index.js:130]: Should `isCanceled()` be directly inlined there?

```suggestion
				isCanceled: context.isCanceled || (parsed.options.signal ? parsed.options.signal.aborted : false);
```

**Review by ehmicky — APPROVED** [review]
This looks good to me, thanks a lot @jopemachine!

Now waiting for another review from @sindresorhus.

File: readme.md
Comment (sindresorhus) [inline readme.md:290]: This needs to be added to the TS docs too.

Comment (sindresorhus) [inline readme.md:290]:
```suggestion
You can cancel the spawned process using the [`signal`](#signal-1) option.
```

File: readme.md
Comment (sindresorhus) [inline readme.md:546]:
```suggestion
Type: [`AbortSignal`](https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal)
```

File: readme.md
Comment (sindresorhus) [inline readme.md:550]: The readme and TS docs should be in sync.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Pull request merged**
